Implemented car registration

Thao-Tran opened the pull request:
After connecting to wifi, the car will create a new socket which binds to its new IP address and sends a registration request to the server. The car will continue to retry forever at 1 minute intervals until it receives an ACK from the server with a `car_id` field. Currently, it doesn't do anything with the received `car_id`.

Thao-Tran committed:
Issue #6 - Implement car registration
* Added a method that sends a message to the server to register the car
* If the car registration fails, it retries forever at 1 minute intervals until
it succeeds

Thao-Tran committed:
Merge branch 'master' into thao/car-registration

Thao-Tran committed:
Fixed issues with car registration implementation
* Changed server address to tuple
* Added socket timeout exception catching

igorveselinovic commented on handlers.py:
```diff
@@ -40,6 +47,37 @@ def handle_get_ssid(server, body, addr):
+    userID -- The ID of the user that is registering the car
+    """
+    server = Server(WLAN_IP, PORT, 60)
+    server.add_handler(MsgType.ACK, None)
```
I think this line is unnecessary.

Thao-Tran replied:
It actually doesn't receive if I don't add a handler because the `receive` function checks if the type is registered in the handlers of the Server instance, so it ends up just sending an ERROR response

Thao-Tran replied:
While I was driving, I was thinking about just changing the receive function to do the basic receive and put the type handling in the `receive_forever`

igorveselinovic replied:
That makes sense to me.

igorveselinovic commented on handlers.py:
```diff
@@ -40,6 +47,37 @@ def handle_get_ssid(server, body, addr):
+    """
+    server = Server(WLAN_IP, PORT, 60)
+    server.add_handler(MsgType.ACK, None)
+    attempts = 0
```
This variable is unused.

igorveselinovic commented on handlers.py:
```diff
@@ -40,6 +47,37 @@ def handle_get_ssid(server, body, addr):
+        # Wait for a max of 1 minute until the ACK from the server is received
+        start = time.time()
+        valid = False
+        while time.time() - start < 60:
```
If the timeout is set to 60 already, then I don't think this loop is necessary.

Thao-Tran replied:
I made the loop in case the server receives some other unwanted packet before the receive times out, so that it'd go back to receiving until the 60 second mark is hit.

igorveselinovic replied:
Ah, I get it now. Now that I think about it, if the central server's response packet gets dropped, the car-pi probably shouldn't wait 60 seconds to retry though.

igorveselinovic replied:
Or if the car-pi's ~~REG_USER~~ REG_CAR packet gets dropped for example, it should only probably wait a couple of seconds before retrying.

Thao-Tran replied:
Okay, I'll just lower the retry time to like 5 seconds then. Yeah 60 seemed like a lot when I put it in originally but then I just forgot about it haha.

igorveselinovic commented on handlers.py:
```diff
@@ -76,3 +114,5 @@ def handle_connect_wifi(server, body, addr):
 
     subprocess.run('./shell-scripts/stop-ap.sh')
+
+    carID = register_car(body['user_id'])
```
We might want to come up with a different way to trigger this, because at this point the original `Server` object is useless since the AP isn't up anymore. Maybe the original server should close itself here and return `user_id` to `main.py` where the `register_car` function could then be called.

Thao-Tran replied:
True, yeah I forgot about the original Server object. That's a good suggestion, I'll see what we can do.

Thao-Tran committed:
Added Metadata and CloseServer classes

Thao-Tran committed:
Fixed issues brought up in code review
* Original Server instance is closed after connecting to WiFi and a new
Server is created to bind to the wlan0 interface instead of the access
point interface
* Car registration now has a timeout of 5 seconds instead of 60
* `user_id` and `car_id` are stored in the metadata file

Thao-Tran committed:
Fixed connect-wifi script

Thao-Tran committed:
Added retry for WIFI_CONN ACK

igorveselinovic commented on handlers.py:
```diff
@@ -72,7 +111,24 @@ def handle_connect_wifi(server, body, addr):
+    # Retry if client sends the WIFI_CONN message within 10 seconds to indicate
+    # that the ACK message was dropped.
+    server.socket.settimeout(10)
+    while True:
```
I think we should limit this to 3 attempts, so that it doesn't end up waiting forever.
